### PR TITLE
block COMPLIANCE to GOVERNANCE retention mode downgrade

### PR DESCRIPTION
**Pull request by kajalpareek-lab**
## Summary
- Reject PutObjectRetention when an object has active COMPLIANCE retention and the request tries to change mode away from COMPLIANCE (e.g. to GOVERNANCE).
- Keep allowing COMPLIANCE date extension, GOVERNANCE → COMPLIANCE upgrade, and new retention after the retain-until date has passed.
- Fix the existing S3 Object Lock test so it uses a non-shorter retain date (otherwise it only exercised "cannot shorten", not mode downgrade), and add unit tests for the retention mode rules.

## Test plan
- [x] Unit tests: `put_object_retention` cases in `test_object_server.test.js` (downgrade blocked, upgrade allowed, extend allowed, expired allowed, shorten blocked)
- [ ] Integration: `test_s3_worm.js` — `should fail to change compliance mode to governance`
- [ ] Manual: PutObjectRetention COMPLIANCE → GOVERNANCE with same/longer date returns AccessDenied; GOVERNANCE → COMPLIANCE still succeeds

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented active COMPLIANCE retention modes from being changed.
  * Preserved protections against shortening or clearing locked retention periods.
  * Improved handling of missing retention requests to avoid errors.
  * Added clearer validation for retention-mode changes.

* **Tests**
  * Added coverage for retention-mode transitions, expiration, date changes, and authorization failures.
  * Verified that governance-locked objects can be upgraded to COMPLIANCE mode successfully.

### CODE REVIEW

File: src/server/object_services/object_server.js [hunk 1]
```diff
@@ -348,15 +348,29 @@ async function put_object_retention(req) {
...
-            info.lock_settings.retention.mode === 'COMPLIANCE') {
-            dbg.error('put object retention failed due object retention mode', obj);
+    const current_retention = info.lock_settings && info.lock_settings.retention;
+    if (current_retention) {
+        const now = new Date();
+        const current_retain_until = new Date(current_retention.retain_until_date);
+        const new_retention = req.rpc_params.retention;
+
+        // AWS: while COMPLIANCE retention is active, mode cannot change (e.g. to GOVERNANCE).
+        // Must run outside the shorten-date check — same/longer dates previously bypassed it.
```
Comment (nadavMiz) [inline src/server/object_services/object_server.js]: may be personal, but I feel like this comment is not very understandable. what does it mean: "Must run outside the shorten-date check — same/longer dates previously bypassed it."?
I feel like it should be more intuitive to someone that doesn't necessarily understand object lock semantics. something like:
"Cannot change the COMPLIANCE mode when replacing an existing policy with a new one"

File: src/server/object_services/object_server.js
Comment (nadavMiz) [inline src/server/object_services/object_server.js:335]: I think its worth adding a small comment explaining the function. and generally what it allows and disallows

File: src/test/integration_tests/api/s3/test_s3_worm.js [hunk 2]
```diff
@@ -745,16 +745,18 @@ mocha.describe('s3 worm', function() {
         });
 
         mocha.it('should fail to change compliance mode to governance', async function() {
-            const futureDate = new Date();
-            futureDate.setDate(futureDate.getDate() + 30);
+            // Use a date >= current retain-until (extended to +60 above).
+            // A shorter date would fail for the wrong reason (cannot shorten COMPLIANCE).
+            const longerOrEqualDate = new Date();
+            longerOrEqualDate.setDate(longerOrEqualDate.getDate() + 90);
```
Comment (nadavMiz) [inline src/test/integration_tests/api/s3/test_s3_worm.js]: why is this change needed? why does it matters if we are 30 or 90 days in the future?
also its never an equal date

File: src/test/unit_tests/internal/test_object_server.test.js [hunk 3]
```diff
@@ -543,3 +543,175 @@ describe('object_server - update_bulk_delete_results', () => {
         expect(results[3]).toHaveProperty('seq', 103);
     });
 });
+
+describe('object_server - put_object_retention', () => {
```
Comment (nadavMiz) [inline src/test/unit_tests/internal/test_object_server.test.js]: I don't dislike the idea of adding unit tests. but we are already using integration tests for object lock. I think it would be good to have this coverage in test_s3_worm.js. the full path doesn't add much complexity, and I feel like its worth testing. we would also get the added benefit of testing NC as well.

having said that if you still feel like unit tests are needed we should probably move them to their own file for object_lock. no one would expect object lock tests to be here, and I fear they would be missed and potentially recreated elsewhere

File: src/server/object_services/object_server.js [hunk 4]
```diff
@@ -348,15 +348,29 @@ async function put_object_retention(req) {
...
+        if (new_retention &&
+            current_retention.mode === 'COMPLIANCE' &&
+            current_retain_until > now &&
+            new_retention.mode !== 'COMPLIANCE') {
+            dbg.error('put object retention failed: cannot change active COMPLIANCE mode', obj);
             throw new RpcError('UNAUTHORIZED');
         }
+
+        if (!new_retention || new Date(new_retention.retain_until_date) < current_retain_until) {
+            if ((current_retention.mode === 'GOVERNANCE' && (!req.rpc_params.bypass_governance || req.role !== 'admin')) ||
```
Comment (nadavMiz) [inline src/server/object_services/object_server.js]: worth adding a TODO that bypass_governance should be allowed to users with IAM permission and not specificly to admin

File: src/test/integration_tests/api/s3/test_s3_worm.js (hunk 2 above)
Comment (nadavMiz) [reply src/test/integration_tests/api/s3/test_s3_worm.js]: ahh its because of the previous test. then its  a good fix. you should just change the name of the variable

File: src/server/object_services/object_server.js (hunk 4 above)
Comment (shirady) [reply src/server/object_services/object_server.js]: @nadavMiz I think the `req.role` admin is not related to the IAM users.
@kajalpareek-lab I don't think you need to check the `req.role` again, as it is already checked in the function before:
```
   if (req.role !== 'admin') {
        throw new RpcError('UNAUTHORIZED');
    }
```

and anyway I think that you can test the feature of object lock with IAM users vs root users to see you have the expected results according to AWS and if not fix accordingly.

File: src/server/object_services/object_server.js [hunk 5]
```diff
@@ -348,15 +348,29 @@ async function put_object_retention(req) {
...
+                current_retention.mode === 'COMPLIANCE') {
+                dbg.error('put object retention failed due object retention mode', obj);
```
Comment (shirady) [inline src/server/object_services/object_server.js]: Could you add the `new_retention` in the logs? something like:

```suggestion
                dbg.error('put object retention failed due object retention mode', obj, 'new_retention', new_retention);
```

Could you share what we see in this log?
I want to see that printing the whole `obj` is not overwhelming (since the interesting details are the object name and the retention mode).

File: src/test/integration_tests/api/s3/test_s3_worm.js (hunk 2 above)
Comment (shirady) [reply src/test/integration_tests/api/s3/test_s3_worm.js]: @kajalpareek-lab could you add a comment about the days so it would be clear for others? (the dependency between the tests).

**Review by shirady — COMMENTED** [review]
I agree with @nadavMiz comments and added a few comments

File: src/server/object_services/object_server.js (hunk 1 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: Updated. Changed it to:
"Cannot change COMPLIANCE mode when replacing an existing retention policy (e.g. COMPLIANCE -> GOVERNANCE) while retention is still active."
Also kept a short note that this check is separate from date shortening so a longer retain-until cannot bypass the mode restriction.

File: src/server/object_services/object_server.js (hunk 4 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: @nadavMiz Added a TODO that BypassGovernanceRetention should be allowed based on IAM / bucket-policy permission, not specifically the admin role.

File: src/test/unit_tests/internal/test_object_server.test.js (hunk 3 above)
Comment (kajalpareek-lab) [reply src/test/unit_tests/internal/test_object_server.test.js]: Agreed. Moved the unit tests out of `test_object_server.test.js`. I’ll also add the COMPLIANCE -> GOVERNANCE / allowed transition coverage to `test_s3_worm.js` so we exercise the full S3 path (including NC).

File: src/server/object_services/object_server.js
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js:335]: Added a short JSDoc on put_object_retention describing what it allows/disallows for COMPLIANCE vs GOVERNANCE while retention is active.

File: src/test/integration_tests/api/s3/test_s3_worm.js (hunk 2 above)
Comment (kajalpareek-lab) [reply src/test/integration_tests/api/s3/test_s3_worm.js]: @nadavMiz Yes — needed because the previous test extends retain-until to +60 days, so this one must use a later date (+90) so the failure is on COMPLIANCE -> GOVERNANCE mode change, not on shortening.
Renamed the variable to `longerRetainUntilDate` and added a comment explaining the dependency between the two tests.

File: src/server/object_services/object_server.js (hunk 4 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: @shirady Removed the redundant admin check in the shorten/bypass path. Admin is already enforced earlier in the function. Agree that IAM-user vs root-user Object Lock coverage is useful; I’d prefer to handle that as a follow-up outside this COMPLIANCE downgrade fix.

File: src/server/object_services/object_server.js (hunk 5 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: Updated the failure logs to include `new_retention`, for example:
`dbg.error('put object retention failed: cannot change active COMPLIANCE mode', obj, 'new_retention', new_retention);`
If logging the full `obj` is too noisy, I can reduce it to key fields (object name / id and current retention). Let me know if you want that in this PR.

File: src/test/integration_tests/api/s3/test_s3_worm.js (hunk 2 above)
Comment (kajalpareek-lab) [reply src/test/integration_tests/api/s3/test_s3_worm.js]: @shirady Added a comment explaining the dependency between the tests: previous test sets +60, so this one uses +90 to assert mode-change denial rather than date shortening.

File: src/server/object_services/object_server.js (hunk 5 above)
Comment (shirady) [reply src/server/object_services/object_server.js]: Could you attach what you see in the logs? so we can see if it's too noisy

Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: Full obj was noisy, so I changed the log to only key / obj_id / current_retention / new_retention. Example below — that should be enough without dumping the whole object_md. 

`Aug-3 14:45:19.056 [/8928] [ERROR] core.server.object_services.object_server:: put object retention failed: cannot change active COMPLIANCE mode { key: 'locked-key', obj_id: 'objidhex', current_retention: { mode: 'COMPLIANCE', retain_until_date: 2026-08-18T09:15:19.055Z }, new_retention: { mode: 'GOVERNANCE', retain_until_date: 2026-08-18T09:15:19.055Z } }`

File: src/server/object_services/object_server.js [hunk 6]
```diff
@@ -331,9 +331,13 @@ async function get_object_retention(req) {
     };
 }
 /**
- *
  * put_object_retention
  *
+ * Sets or updates object retention (mode + retain-until date) when Object Lock is enabled.
+ * While retention is active:
+ * - COMPLIANCE: cannot shorten retain-until, and cannot change mode (e.g. to GOVERNANCE).
+ * - GOVERNANCE: cannot shorten retain-until unless bypass_governance is set.
```
Comment (shirady) [inline src/server/object_services/object_server.js]: Can we change the mode from GOVERNANCE -> GOVERNANCE?
It is implicit from the JSDoc
If so, could you write this explicitly?

File: src/server/object_services/object_server.js [hunk 7]
```diff
@@ -343,21 +347,50 @@ async function put_object_retention(req) {
     const obj = await find_object_md(req);
     const info = get_object_info(obj, { role: req.role });
 
+    // TODO: BypassGovernanceRetention should be allowed for callers with IAM /
+    // bucket-policy permission, not only the admin role.
```
Comment (shirady) [inline src/server/object_services/object_server.js]: This TODO is not related to the check `if (req.role !== 'admin')`; this is the `auth` of the `RPC`.

File: src/server/object_services/object_server.js [hunk 8]
```diff
@@ -343,21 +347,50 @@ async function put_object_retention(req) {
...
+        const new_retention = req.rpc_params.retention;
+
+        // Cannot change COMPLIANCE mode when replacing an existing retention policy
+        // (e.g. COMPLIANCE -> GOVERNANCE) while retention is still active.
+        // Checked separately from date shortening so same/longer retain-until dates cannot bypass it.
+        if (new_retention &&
+            current_retention.mode === 'COMPLIANCE' &&
+            current_retain_until > now &&
+            new_retention.mode !== 'COMPLIANCE') {
+            // Log only retention-relevant fields; full object_md is too noisy.
```
Comment (shirady) [inline src/server/object_services/object_server.js]: I don't think we need this comment

File: src/server/object_services/object_server.js [hunk 9]
```diff
@@ -343,21 +347,50 @@ async function put_object_retention(req) {
...
+            dbg.error('put object retention failed: cannot change active COMPLIANCE mode', {
+                key: obj.key,
+                obj_id: obj._id && (obj._id.toHexString ? obj._id.toHexString() : String(obj._id)),
```
Comment (shirady) [inline src/server/object_services/object_server.js]: Why do you need the object_id in the logs?
Why use the `obj` when you have the `info` (from `get_object_info`) and you can easily use: `info.obj_id`?

File: src/server/object_services/object_server.js [hunk 10]
```diff
@@ -343,21 +347,50 @@ async function put_object_retention(req) {
...
+                current_retention,
+                new_retention,
+            });
             throw new RpcError('UNAUTHORIZED');
         }
+
+        if (!new_retention || new Date(new_retention.retain_until_date) < current_retain_until) {
+            // req.role === 'admin' was already enforced above.
```
Comment (shirady) [inline src/server/object_services/object_server.js]: I don't think we need that comment

File: src/server/object_services/object_server.js [hunk 11]
```diff
@@ -343,21 +347,50 @@ async function put_object_retention(req) {
...
+            if ((current_retention.mode === 'GOVERNANCE' && !req.rpc_params.bypass_governance) ||
+                current_retention.mode === 'COMPLIANCE') {
+                dbg.error('put object retention failed due object retention mode', {
+                    key: obj.key,
+                    obj_id: obj._id && (obj._id.toHexString ? obj._id.toHexString() : String(obj._id)),
+                    current_retention,
+                    new_retention,
+                });
+                throw new RpcError('UNAUTHORIZED');
```
Comment (shirady) [inline src/server/object_services/object_server.js]: Could you check the error message in AWS for those cases to see if we have the same error message?

File: src/test/unit_tests/internal/test_object_lock_retention.test.js
Comment (shirady) [inline src/test/unit_tests/internal/test_object_lock_retention.test.js:1]: What are this new tests add that we do not cover in `src/test/unit_tests/internal/test_object_lock_retention.test.js`?

File: src/server/object_services/object_server.js (hunk 6 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: Yes — same-mode updates are allowed when retain-until is the same or later. Updated the JSDoc to call out COMPLIANCE → COMPLIANCE and GOVERNANCE → GOVERNANCE explicitly.

File: src/server/object_services/object_server.js (hunk 7 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: Agree — that TODO is about BypassGovernanceRetention permission, not RPC admin auth. Moved it next to the GOVERNANCE shorten / bypass check.

File: src/server/object_services/object_server.js (hunk 8 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: Removed.

File: src/server/object_services/object_server.js (hunk 9 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: Keeping obj_id so we can correlate retention denials with a specific object version in support logs (key alone isn’t enough when there are multiple versions). Switched to info.obj_id from get_object_info.

File: src/server/object_services/object_server.js (hunk 10 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: removed.

File: src/test/unit_tests/internal/test_object_lock_retention.test.js
Comment (kajalpareek-lab) [reply src/test/unit_tests/internal/test_object_lock_retention.test.js:1]: The unit file keeps the edge cases that are awkward or slow over S3: same-date downgrade (proves mode check isn’t only hit via shorten), longer-date downgrade, expired COMPLIANCE then GOVERNANCE, and isolated put_object_retention without a cluster. Per Nadav’s earlier note we moved them out of test_object_server.test.js into this dedicated file.

File: src/server/object_services/object_server.js (hunk 11 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: Checked on a live hosted cluster with this change: PutObjectRetention denials (COMPLIANCE → GOVERNANCE and shortening COMPLIANCE) return AccessDenied / Access Denied, same as other Object Lock denials and what our worm tests already expect for hosted. That matches AWS’s AccessDenied behavior for these cases. I’d leave the message as-is in this PR.

File: src/server/object_services/object_server.js [hunk 12]
```diff
@@ -349,15 +354,42 @@ async function put_object_retention(req) {
...
+                obj_id: info.obj_id,
+                current_retention,
+                new_retention,
+            });
             throw new RpcError('UNAUTHORIZED');
         }
+
+        if (!new_retention || new Date(new_retention.retain_until_date) < current_retain_until) {
+            // TODO: BypassGovernanceRetention should follow IAM / bucket-policy permission
+            // (not only callers that already passed RPC admin auth above).
```
Comment (shirady) [inline src/server/object_services/object_server.js]: It is not related to this code block, right?
The bucket policy and IAM user inline policy were authorized in `rest_s3.
Could you remove the comment?

File: src/test/unit_tests/internal/test_object_lock_retention.test.js [hunk 13]
```diff
@@ -0,0 +1,181 @@
+/* Copyright (C) 2026 NooBaa */
+
+'use strict';
+
+const CONSTANTS = require('../../../common/constants');
```
Comment (shirady) [inline src/test/unit_tests/internal/test_object_lock_retention.test.js]: Can you destruct and only have the VERSIONING so you can use it?

File: src/test/unit_tests/internal/test_object_lock_retention.test.js [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+        const req = make_req({
+            mode: 'GOVERNANCE',
+            retain_until_date: new_until,
+        });
+
+        await expect(object_server.put_object_retention(req)).resolves.toBeUndefined();
+        expect(update_object_by_id_stub).toHaveBeenCalled();
+    });
+
+    test('still rejects shortening active COMPLIANCE retention', async () => {
```
Comment (shirady) [inline src/test/unit_tests/internal/test_object_lock_retention.test.js]: Why is the name with "still"?
```suggestion
    test('rejects shortening active COMPLIANCE retention', async () => {
```

File: src/server/object_services/object_server.js [hunk 15]
```diff
@@ -331,9 +331,14 @@ async function get_object_retention(req) {
...
  * put_object_retention
  *
+ * Sets or updates object retention (mode + retain-until date) when Object Lock is enabled.
+ * While retention is active:
+ * - COMPLIANCE: cannot shorten retain-until, and cannot change mode (e.g. to GOVERNANCE).
+ * - GOVERNANCE: cannot shorten retain-until unless bypass_governance is set.
+ * - Same mode is allowed: COMPLIANCE -> COMPLIANCE and GOVERNANCE -> GOVERNANCE
+ *   when retain-until is the same or later (extend / no-op date).
  */
 async function put_object_retention(req) {
```
Comment (shirady) [inline src/server/object_services/object_server.js:337]: BTW, with all the changes, the function became quite long, if you can reduce the number of lines.

File: src/server/object_services/object_server.js (hunk 7 above)
Comment (shirady) [reply src/server/object_services/object_server.js]: Why do you need to move it?
When you add the fix, would it be in the `object_server`?
I think you can remove it from the code, and if you want, you can document it in the PR description.

File: src/server/object_services/object_server.js (hunk 11 above)
Comment (shirady) [reply src/server/object_services/object_server.js]: I meant the error message, not just the error code.

File: src/server/object_services/object_server.js (hunk 7 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: Removed.

File: src/server/object_services/object_server.js (hunk 15 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js:337]: Extracted _throw_if_retention_update_forbidden so put_object_retention is the auth/load/update path only. Behavior unchanged validation lives in one helper.

File: src/test/unit_tests/internal/test_object_lock_retention.test.js (hunk 13 above)
Comment (kajalpareek-lab) [reply src/test/unit_tests/internal/test_object_lock_retention.test.js]: Done

File: src/test/unit_tests/internal/test_object_lock_retention.test.js (hunk 14 above)
Comment (kajalpareek-lab) [reply src/test/unit_tests/internal/test_object_lock_retention.test.js]: Renamed.

File: src/server/object_services/object_server.js (hunk 12 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: Removed. BypassGovernanceRetention IAM/bucket-policy auth belongs in s3_rest / #9881, not in this retention validation path.

File: src/server/object_services/object_server.js (hunk 11 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: You’re right — I was looking at the error code only.
Hosted today still returns "Access Denied" for these PutObjectRetention denials (UNAUTHORIZED). Aligning the message to "Access Denied because object protected by object lock." (and mapping via OBJECT_LOCKED / AccessDeniedObjectLocked) is already covered in #9868, including the worm test updates for both hosted and NC.
I’ll leave the message as-is in this PR and rebase onto #9868 after it merges.

**Review by nadavMiz — APPROVED** [review]

**Comment by nadavMiz** [comment]
looks like error message is still wrong. after that is fixed. its fine by me

**Comment by shirady** [comment]
> looks like error message is still wrong. after that is fixed. its fine by me

@nadavMiz @kajalpareek-lab plans to fix it in another PR - see [comment](https://github.com/noobaa/noobaa-core/pull/9823#discussion_r3763169012).

File: src/server/object_services/object_server.js [hunk 16]
```diff
@@ -349,33 +354,71 @@ async function put_object_retention(req) {
...
+        key: obj.key,
+        obj_id: info.obj_id,
+        current_retention,
+        new_retention,
+        bypass_governance: Boolean(req.rpc_params.bypass_governance),
+    });
+
+    const legal_hold = info.lock_settings?.legal_hold ?
+        { status: info.lock_settings.legal_hold.status } :
+        undefined;
```
Comment (shirady) [inline src/server/object_services/object_server.js]: Why do you need a ternary conditional operator?
I think with the optional chining that you're using, if we don't have the value, the `legal_hold` value will be `undefined`, so you can continue the optional chaining for `status`, unless I'm missing something...

```suggestion
    const legal_hold = info.lock_settings?.legal_hold?.status
```

File: src/server/object_services/object_server.js [hunk 17]
```diff
@@ -349,33 +354,71 @@ async function put_object_retention(req) {
...
+/**
+ * Enforce Object Lock retention update rules for an existing active retention.
+ */
+function _throw_if_retention_update_forbidden({
+    key,
+    obj_id,
+    current_retention,
+    new_retention,
+    bypass_governance,
+}) {
```
Comment (shirady) [inline src/server/object_services/object_server.js]: I think that without defaulting the object when you destruct, calling the function without an object is prone to the error `TypeError`.
I think you can add it:

```suggestion
function _throw_if_retention_update_forbidden({
    key,
    obj_id,
    current_retention,
    new_retention,
    bypass_governance,
} = {}) {
```

File: src/server/object_services/object_server.js [hunk 18]
```diff
@@ -2688,6 +2731,6 @@ exports.update_transition_status = update_transition_status;
 
 if (process.env.NODE_ENV === 'test') {
     exports.__testing = {
-        update_bulk_delete_results
+        update_bulk_delete_results,
```
Comment (shirady) [inline src/server/object_services/object_server.js:2790]: No change here, right?
So you can remove this addition (fewer changed lines).

File: src/test/integration_tests/api/s3/test_s3_worm.js [hunk 19]
```diff
@@ -745,20 +745,58 @@ mocha.describe('s3 worm', function() {
         });
 
         mocha.it('should fail to change compliance mode to governance', async function() {
-            const futureDate = new Date();
-            futureDate.setDate(futureDate.getDate() + 30);
+            // Previous test extended COMPLIANCE retain-until to +60 days.
+            // Use +90 here (after that date) so this assert fails on mode change
+            // (COMPLIANCE -> GOVERNANCE), not on "cannot shorten COMPLIANCE".
+            const longerRetainUntilDate = new Date();
```
Comment (shirady) [inline src/test/integration_tests/api/s3/test_s3_worm.js]: Our code style is snake case:
```suggestion
            const longer_retain_until_date = new Date();
```

File: src/test/unit_tests/internal/test_object_lock_retention.test.js
Comment (shirady) [reply src/test/unit_tests/internal/test_object_lock_retention.test.js:1]: @nadavMiz by using this file it would not be tested in NC, could you explain why not adding the test cases in `src/test/integration_tests/api/s3/test_s3_worm.js`?

I'm trying to look at the test cases, and I think it is good coverage for both environments:
- `rejects COMPLIANCE to GOVERNANCE downgrade with same retain date`
- `rejects COMPLIANCE to GOVERNANCE downgrade with longer retain date`
- `'allows GOVERNANCE to COMPLIANCE upgrade with same retain date`
- `allows extending COMPLIANCE retention while keeping COMPLIANCE mode`
- `allows setting GOVERNANCE after COMPLIANCE retention has expired`
- `rejects shortening active COMPLIANCE retention`

File: src/server/object_services/object_server.js (hunk 16 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: removed the old ternary form and now use optional chaining for the status read, then build the legal_hold object from that value.

File: src/server/object_services/object_server.js (hunk 17 above)
Comment (kajalpareek-lab) [reply src/server/object_services/object_server.js]: Agreed. Added a default object in the helper signature (= {}) so accidental direct calls won’t throw on destructuring.

File: src/test/integration_tests/api/s3/test_s3_worm.js (hunk 19 above)
Comment (kajalpareek-lab) [reply src/test/integration_tests/api/s3/test_s3_worm.js]: Agreed. Renamed to snake_case: + "longer_retain_until_date" + .
